New Ledger API test tool

stefanobaghino-da opened the pull request:
### To begin with

The new app represents a rough prototype of the new ledger _acceptance_ test tool, aimed at verifying a ledger implementation (even ours) against a set of tests.

### To run

    bazel run //ledger/ledger-api-test-tool:ledger-acceptance

As of https://github.com/digital-asset/daml/commit/1db882b01f0ef21b4f35d24a44680df224315b2d:
- only the ledger identity test is supposed to work
- the tests are only ran against a managed sandbox, not setup required so far

### To understand

A few notes on the most important classes (as of https://github.com/digital-asset/daml/commit/1db882b01f0ef21b4f35d24a44680df224315b2d):

- `LedgerSession`: represents a connection to a given ledger
  - the life cycle of the ledger right now can only be _managed_ (i.e. in process) but the next step is to provide a way to connect to a _unmanaged_ ledger
  - right now many helpers are implemented here but they should probably be moved into their own class and then accessed from the test suites from there
- `LedgerTestSuite`: represents a set of logically bound tests, it's passed all the `LedgerSession`s it should run against at runtime
- `LedgerTest`: it represents a single test -- for each test a `LedgerTestContext` is passed
- `LedgerTestContext`: provides access to an execution context and some test-specific information (e.g. the ledger end offset when the test started) -- test-specific helpers are also implemented here
- `LedgerTestSuiteRunner`: as the name suggests, it takes a test suite and runs it on an execution context, producing results in the form of `LedgerTestSummary` objects 
- `LedgerTestSummary`: information about a test run, including its `Result`, that can be passed to a `Reporter` for rendering
- `Main`: it should probably become a separate class able to stitch together ledger connections and tests to be run, some sort of supervisor for multiple the `LedgerTestSuiteRunner`

### To review
- ~the `LedgerClock` is simply broken and only works when falling back on wall clock time, I can use some help in understanding what I'm doing wrong there~
- as of https://github.com/digital-asset/daml/commit/1db882b01f0ef21b4f35d24a44680df224315b2d, apart from the general architecture and implementation, I would like to understand how the helpers in `LedgerTestSuite` (via `LedgerSession` and `LedgerTestContext`) can be implemented in terms of our own bindings
- given what I've seen in the existing integration tests, it may be the case that we may want to still run some tests as integration tests with the old framework (e.g.: testing that a _sandbox_ can be started with a given ledger identifier -- which is not a property of arbitrary ledger implementations) 

### To do

#### Checklist as of https://github.com/digital-asset/daml/commit/1db882b01f0ef21b4f35d24a44680df224315b2d

- [x] don't use the global fork-join pool as the hard-coded execution context for tests
- [x] make the application return an error code when any test fails
- [x] extract helpers from the `LedgerSession` into their own class
- [ ] ~design and implement a way to run against multiple participants in the same test~
- [x] extract everything related to running tests from the `Main`
- [x] make the new app a drop-in replacement to the existing API test tool (i.e. use its CLI)
- [x] offer a way to connect with a non-managed ledger (i.e. ran out of process)
- [ ] ~determine dependencies between tests and linearize their implementation†~
- [ ] ~add all relevant tests to this tool~
- [ ] ~evaluate adding more tests when and where needed~
- [ ] ~add alternative reporters (e.g. JUnit-compatible reporter)~

---

† This could probably be easily translated into compatibility tiers for DAML ledger implementors

---

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

gerolf-da commented on ledger/ledger-api-test-tool/BUILD.bazel:
```diff
@@ -84,6 +86,35 @@ da_scala_binary(
+        "src/main/resources/logback.xml",
+    ],
+    tags = [
+        "maven_coordinates=com.daml.ledger.testtool:ledger-api-test-tool:__VERSION__",
```
Did you want to use the same maven coordinates because the new tool will eventually act as a drop-in replacement for the existing test tool?

stefanobaghino-da replied:
No, it's a whole lotta copy :spaghetti:.

stefanobaghino-da replied:
I'm adopting everything from the test tool, so the answer to your question turned to a yes.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala:
```diff
@@ -0,0 +1,274 @@
+      templateIds: Identifier*)(implicit context: ExecutionContext): Future[Vector[Transaction]] =
+    for {
+      id <- ledgerId()
+      end <- ledgerEnd()
```
Technically you don't need to fetch the ledger end first. You can use the [LedgerBoundary#LEDGER_END](https://github.com/digital-asset/daml/blob/master/ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/ledger_offset.proto#L31) instead and the ledger will do the "what's the current ledger end" check for you.

stefanobaghino-da replied:
Fixed by https://github.com/digital-asset/daml/pull/2282/commits/577409f56cb48d68561a79521e4abac9db92cc59

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala:
```diff
@@ -0,0 +1,274 @@
+        new SubmitAndWaitRequest(
+          Some(new Commands(
+            ledgerId = id,
+            applicationId = UUID.randomUUID().toString,
```
Keep in mind that you can only receive completions for known applicationIds.

stefanobaghino-da replied:
Each test case will have its own, automatically generated application identifier starting from https://github.com/digital-asset/daml/pull/2282/commits/d8efe8282cb58b5e7bacb4a4b2792e30b848be9e

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala:
```diff
@@ -0,0 +1,274 @@
+      inFiveSeconds = now.plusSeconds(5)
+      tx <- commandService.submitAndWaitForTransaction(
+        new SubmitAndWaitRequest(
+          Some(new Commands(
```
At some point we might want to use `workflowId` as well and expose it as parameter.

stefanobaghino-da replied:
I'll keep it simple for now, I believe that one application identifier per test case right now is all we need, but I'll definitely introduce this at a later time if we need a more granular control. Probably I'll let the `LedgerTestContext` provide `workflowId`s that can then in turn be used to filter transactions into more fine grained workflows in a test.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala:
```diff
@@ -0,0 +1,274 @@
+              ledgerEffectiveTime = Some(new Timestamp(now.getEpochSecond, now.getNano)),
+              maximumRecordTime =
+                Some(new Timestamp(inFiveSeconds.getEpochSecond, inFiveSeconds.getNano)),
+              commands = Seq(
```
Considering that `commands` is the only difference between `create` and `exercise`, maybe we can deduplicate the fluff around it.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/5094794b8fe366b4f1ba082f1fb87a8190672797

This commit also adds helpers specific for test DAML code.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/tests/Divulgence.scala:
```diff
@@ -0,0 +1,67 @@
+          assert(
+            contractId == divulgence2,
+            "The only visible event should be the creation of the second contract")
+        }
```
This test is not yet complete, it's missing further assertions: https://github.com/digital-asset/daml/blob/2a17b3d1911e6b62dcb343ebf9c5aefdcceef657/ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/DivulgenceIT.scala#L284-L323

stefanobaghino-da commented:
Current status:
- semantic tests have been integrated
- implementation of divulgence tests [still needs to be completed](https://github.com/digital-asset/daml/pull/2282#discussion_r308084975) (it's the next step, alongside other tests)
- most of the CLI options have been adapted to be seamlessly interpreted by the new tool
  - exceptions are `--stable-party-identifiers` and `--stable-command-identifiers`, which could represent a problem given the current design -- @mziolekda @dajmaki how essential are those flags?
- open points from the initial draft that i would leave for a later PR
  - design and implement a way to run against multiple participants in the same test
    - I would prefer to have a concrete case before designing a framework around it
  - determine dependencies between tests and linearize their implementation
    - I would abandon this idea temporarily but I still see value in this for implementors
  - add all relevant tests to this tool
    - This can be done one by one in subsequent PRs
  - evaluate adding more tests when and where needed
    - I would start from the ones that are already available on the existing tool
  - add alternative reporters (e.g. JUnit-compatible reporter)
    - Should be relatively easy but it's low priority
- points that I would like to tackle before merging this
  - [ ] heavily comment the implementation
  - [ ] make a quick write-up about the test tool design
  - [ ] add indications on how to write tests using this framework
  - [ ] swap in the new tool for the old
  - [ ] delete integration tests that are implemented in the new test tool

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/infrastructure/LedgerServices.scala:
```diff
@@ -0,0 +1,24 @@
+import io.grpc.Channel
+
+private[infrastructure] final class LedgerServices(channel: Channel) {
+  val cmd: CommandService = CommandServiceGrpc.stub(channel)
```
Do the abbreviations really improve readability?

stefanobaghino-da replied:
Absolutely not, I can have longer names. :slightly_smiling_face:

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/0ec23c48accfc125699f2dba0b237b4ec81483aa

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/infrastructure/LedgerTestContext.scala:
```diff
@@ -0,0 +1,58 @@
+
+  val ledgerId: Future[String] = bindings.ledgerId
+
+  val offsetAtStart: Future[LedgerOffset] = bindings.ledgerEnd
```
I think this probably should be passed in as a constructor parameter as a non-`Future` value. This has the smell of potential flakiness.

stefanobaghino-da replied:
I see your point, I'll investigate the point further.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/e41532036369394d18a9a5278323f476d24fce77

Thanks for the tip given offline.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/infrastructure/SemanticTesterLedger.scala:
```diff
@@ -0,0 +1,78 @@
+    for {
+      apiCommands <- lfCommandToApiCommand(party, lfCommands)
+      command +: commands = apiCommands.commands.map(_.command)
+      id <- bindings.submitAndWaitForTransactionId(
```
You could change it to `submitAndWaitForTransactionTree`, although it'll have slightly different semantics than the current implementation, which loads all events visible to all `parties`.

stefanobaghino-da replied:
As discussed offline, the semantic tests need broader visibility.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/templates/Divulgence2.scala:
```diff
@@ -0,0 +1,40 @@
+
+}
+
+sealed abstract case class Divulgence2(
```
Why is it an abstract case class?

stefanobaghino-da replied:
Controlling the constructor while letting the compiler synthesize the extractors ([reference](https://gist.github.com/tpolecat/a5cb0dc9adeacc93f846835ed21c92d2)).

gerolf-da replied:
Thanks for the reference 👍

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/templates/Divulgence2.scala:
```diff
@@ -0,0 +1,40 @@
+    contractId: String,
+    party: String,
+    div2Signatory: String,
+    div2Fetcher: String) {
```
This field doesn't seem to be used anywhere.

stefanobaghino-da replied:
Yeah, not sure whether this is used in some other test, I can remove it without consequences for now I guess. Ideally I'd like to supplant this with code generation.

gerolf-da replied:
I guess we'll wait with that until the Divulgence test is complete.

stefanobaghino-da replied:
Yes, I've just swapped in the new tool, I'll finish `DivulgenceIT` and `ContractKeysIT` (since it's used for conformance tests of the reference DAML-on-X implmentation) and then I'd say we can merge, module further review comments, mostly because this PR is getting huge with the swap and subsequent deduplication. Open points can be addressed one by one in upcoming PRs.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/c7e6538d4efaacb3ed283c726acc0a12dfaaa80f

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/tests/Time.scala:
```diff
@@ -0,0 +1,33 @@
+final class Time(session: LedgerSession) extends LedgerTestSuite(session) {
+
+  val pass =
+    LedgerTest("Advancing time should return the new time") { implicit context =>
```
Is this test missing a `skipIf` clause for when the time service is not available?

stefanobaghino-da replied:
Kinda. The problem is that I haven't found a way to express it reliably. Basically right now any option I could introduce would more ore less be equivalent to explicitly `--include=Time` from the CLI.

stefanobaghino-da replied:
Superseded by https://github.com/digital-asset/daml/pull/2282#discussion_r311404661

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/Cli.scala:
```diff
@@ -0,0 +1,130 @@
+
+import com.digitalasset.ledger.api.tls.TlsConfiguration
+
+object Cli {
```
Yes, I initially copied and pasted the code to avoid merge errors. The work I'm doing right now consists in swapping the two implementations, which luckily had no conflicts.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/03dd37dd168713526d000581eec659ce8767e734 (WARNING: it's a **large** commit)

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/Config.scala:
```diff
@@ -0,0 +1,46 @@
+
+import com.digitalasset.ledger.api.tls.TlsConfiguration
+
+final case class Config(
```
Yes, I initially copied and pasted the code to avoid merge errors. The work I'm doing right now consists in swapping the two implementations, which luckily had no conflicts.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/03dd37dd168713526d000581eec659ce8767e734 (WARNING: it's a **large** commit)

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/LedgerApiTestTool.scala:
```diff
@@ -17,25 +17,49 @@ object LedgerApiTestTool {
+      sys.exit(0)
+    }
+
+    val included =
```
I absolutely agree with all the points you made. :wink:

stefanobaghino-da commented:
Just pushed https://github.com/digital-asset/daml/pull/2282/commits/03dd37dd168713526d000581eec659ce8767e734 (WARNING: it's a **large** commit) that:
- swaps in the new test tool
- moves shared DAML test code to `//ledger/test-common` to reduce duplication
- moves facilities for integration tests from `//ledger/test-common` to `//ledger-api/test-utils` (it's probably the case that all these facilities are used exclusively by `//ledger/ledger-api-integration-tests` and can be moved there at a later moment)
- remove semantic tests from `//ledger/ledger-api-integration-tests`

Open points that will be addressed before considering this PR ready to be reviewed and merged:
- [ ] complete divulgence tests
- [ ] implement contract keys tests in test tool
- [ ] remove contract keys tests from integration tests
- [ ] run the test tool on the in-memory sandbox
- [ ] run the test tool on the PostgreSQL-backed sandbox

Striked out points from https://github.com/digital-asset/daml/pull/2282#issuecomment-518267849 will be tackled in subsequent PRs to not make this one too large as addressing the above points puts us in a situation of effective feature parity for external users.

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/infrastructure/LedgerBindings.scala:
```diff
@@ -0,0 +1,210 @@
+    } yield response.ledgerId
+
+  private[this] val clock: Future[LedgerClock] =
+    for (id <- ledgerId; clock <- LedgerClock(id, services.time)) yield clock
```
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/034eb510e6cf739a9fd7ca37570863068e0eca5b

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/rewrite/testtool/infrastructure/LedgerTestSuite.scala:
```diff
@@ -24,7 +24,7 @@ private[testtool] abstract class LedgerTestSuite(val session: LedgerSession) {
   val tests: Vector[LedgerTest] = Vector.empty
 
-  final def skip(reason: String): Future[Unit] = Future.failed(new SkipTestException(reason))
+  final def skip(reason: String): Future[Unit] = throw new SkipTestException(reason)
```
I realized that I was originally actually starting tests from the main thread, which could break the seal around exceptions thrown in the test body (hence all the _PLEASE REPORT THIS_ error messages to cover uncaught exceptions). So I moved to this other approach. I just noticed that now the return type is completely wrong. I could keep this as it was, the most important thing for me was not letting any exception escape the test runner and a call to `skip` made me notice it.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/e41532036369394d18a9a5278323f476d24fce77

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/ApiScenarioTransform.scala:
```diff
@@ -0,0 +1,172 @@
+  Array(
+    "org.wartremover.warts.Any"
+  ))
+class ApiScenarioTransform(ledgerId: String, packages: Map[Ref.PackageId, Ast.Package])
```
Is the intention to remove one of the copies of this file?

stefanobaghino-da replied:
The copy has been already remove from the integration tests because the semantic tests have been already ~eradicated~ removed from there as well.

stefanobaghino-da replied:
Uh, what I just said is wrong. Removing the original as of know. Good catch, thanks!

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/af4baca1712a9dc00000b388d78a82f6f755d6fa

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/templates/Divulgence1.scala:
```diff
@@ -0,0 +1,24 @@
+
+sealed abstract case class Divulgence1(contractId: String, issuer: String, div1Party: String) {
+
+  println(List.fill(3)("#" * 80).mkString("\n"))
```
that's a looooooooooooooooooooooooooooong string of hashes 😄

stefanobaghino-da replied:
Debug prints removed. :wink:

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Time.scala:
```diff
@@ -0,0 +1,29 @@
+    LedgerTest("Advancing time should return the new time") { implicit context =>
+      for {
+        t1 <- time()
+        _ <- passTime(1.second)
```
How does this test work for wall-clock time? Or is it missing a `skipIf`?

stefanobaghino-da replied:
It would throw. The problem is it's quite difficult to reliably detect if the time service is there. In the case of the DAML-on-X reference implementation, for example, the `TimeService` exists also on wall-clock time but it's not capable of doing time travel (which makes sense). So using `skipIf` would not work very reliably unless we signal the compliance with the `TimeService` via configuration/CLI, which could be effectively equivalent to simply specifying to include or exclude the `TimeIT` test. Would that be a feasible approach to the problem?

stefanobaghino-da replied:
You can see an example of at the commit where I introduced running the conformance tests on the in-memory sandbox (https://github.com/digital-asset/daml/pull/2282/commits/9cec58d6d8ab8f83593307f394d3a5d813f8e4a7).

stefanobaghino-da replied:
You can see an example of at the commit where I introduced running the conformance tests on the in-memory (https://github.com/digital-asset/daml/pull/2282/commits/9cec58d6d8ab8f83593307f394d3a5d813f8e4a7) and persistent sandbox (https://github.com/digital-asset/daml/pull/2282/commits/549e26411bd5e7dee2af74a7052c327666de2832).

gerolf-da replied:
Okay, then let's keep this test as optional and running that test should be a conscious opt-in.

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala:
```diff
@@ -0,0 +1,133 @@
+          assert(
+            aliceTransactions.head.events.head.event.created.get.contractId == divulgence1.contractId)
+
+          assert(aliceTransactions.head.events.head.event.created.get.witnessParties == Seq(alice))
```
The original test used both parties to request the transactions for this part of the test, to show that bob does not see this create event.
https://github.com/digital-asset/daml/blob/4ff9a6b7eae9e1ae55c5811f7c8c6f4a5a4d25c1/ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/DivulgenceIT.scala#L321

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/c7e6538d4efaacb3ed283c726acc0a12dfaaa80f

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala:
```diff
@@ -0,0 +1,133 @@
+          assert(
+            activeForAlice.map(_.contractId) == Seq(divulgence1.contractId, divulgence2.contractId))
+          assert(activeForAlice(0).witnessParties == Seq(alice))
+          assert(activeForAlice(1).witnessParties == Seq(alice))
```
Same comment as above. Let's preserve the behavior of the test with respect to both bob and alice requesting transactions.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/c7e6538d4efaacb3ed283c726acc0a12dfaaa80f

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala:
```diff
@@ -0,0 +1,133 @@
+import com.daml.ledger.api.testtool.infrastructure.{LedgerSession, LedgerTest, LedgerTestSuite}
+import com.daml.ledger.api.testtool.templates.{Divulgence1, Divulgence2}
+
+final class Divulgence(session: LedgerSession) extends LedgerTestSuite(session) {
```
`DivulgenceIT` is "decorated" with quite a few comments. Please carry them over to this test as well. I think they are a great help as a preview of what the next few lines are about.
Having said that, the assertions in this test are super helpful as well.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/c7e6538d4efaacb3ed283c726acc0a12dfaaa80f

stefanobaghino-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Divulgence.scala:
```diff
@@ -0,0 +1,133 @@
+          val createDivulgence2Transaction = bobTrees(0)
+          assert(
+            createDivulgence2Transaction.rootEventIds.size == 1,
+            s"The transaction where Divulgence2 is created should contain exactly one root event contains ${createDivulgence2Transaction.rootEventIds.size} instead"
```
I think it's perfect. :slightly_smiling_face:

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/HeadObserver.scala:
```diff
@@ -0,0 +1,42 @@
+
+import scala.concurrent.{Future, Promise}
+
+object HeadObserver {
```
How is this different than `FirstElementObserver` and can we get rid of one of them?

stefanobaghino-da replied:
This cancels the call after we received the first item, so I believe this should not leak open streams. I preferred to duplicate it to not introduce a dependency between the test tool and the integration tests. We can put it in a shared library and get rid of the `FirstElementObserver`. Sounds good?

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/2282/commits/55bd945701279f2b4c8a83f6adf3b57a7f67d983

Renamed to `SingleItemObserver`, made it either get the very first item of the stream or the first that satisfies a predicate, moved it to `//ledger/test-common`, deleted `FirstElementObserver`, replaced all usages. This should be it. :sweat_smile:

stefanobaghino-da commented:
/azp run

S11001001 commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/Reporter.scala:
```diff
@@ -0,0 +1,89 @@
+
+import scala.util.Try
+
+trait Reporter[A] extends ((Vector[LedgerTestSummary]) => A)
```
Is the intent for this to be a newtype for `=>`?

gerolf-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestSuite.scala:
```diff
@@ -0,0 +1,96 @@
+      implicit ec: ExecutionContext): Unit = {
+    assert(
+      t.isInstanceOf[StatusRuntimeException] || t.isInstanceOf[StatusException],
+      "Exception is neither a StatusRuntimeException nor a StatusException")
```
Addressed in #2495.
